### PR TITLE
Port #4928 to release/5.0.2xx branch

### PR DESCRIPTION
**Pull request by mavasani**
### Related Issues:

[dotnet build doesn't terminate in 5.0.200](https://github.com/dotnet/roslyn-analyzers/issues/4920)
[C# compiler freezes in certain platform analysis scenarios](https://github.com/dotnet/roslyn/issues/51652)

### Issue:  
In .NET 5.0.202 (for some corner cases) the build does not terminate , resulting in VS freeze.

### Description: 
The DFA for Platform Compatibility Analyzer enters an infinite loop in certain edge case code patterns, where platform checks are performed inside/outside a loop and there is a try/catch within the loop.  This untested edge case scenario was hidden under [another bug](https://github.com/dotnet/roslyn-analyzers/issues/4372) that when [fixed in 5.0.2](https://github.com/dotnet/roslyn-analyzers/pull/4393) revealed this issue.

### How found:
Customers reported this bug.

### Customer Impact:
All .NET 5.0 SDK (v5.0.200) and above customers. 
Since 5.0.2 release 2 customers reported hitting this issue. As this analyzer is enabled by default with a build warning level, it runs as part of every build with .NET5 SDK.

### Test coverage:
Unit testing and manual testing with console app having same repro.
 
### Risk: 
Low risk. The PR adds couple of bail out checks in DFA merge algorithm which prevents infinite analysis loop. This will also likely improve the overall performance of the analysis in other cases

CC @jaredpar @jeffhandley @vatsalyaagrawal @terrajobst @mikadumont @jmarolf

### CODE REVIEW

**Comment by mavasani** [comment]
@buyaa-n This is the port PR to release/5.0.2xx branch

**Review by jeffhandley — APPROVED** [review]

**Comment by mavasani** [comment]
@buyaa-n I'll let you drive this PR forward.

**Comment by buyaa-n** [comment]
An issue created for VS servicing created, waiting for approval @jmarolf @vatsalyaagrawal please approve the PR when it the servicing approved

**Review by buyaa-n — APPROVED** [review]

**Comment by buyaa-n** [comment]
We have got .NET Core Tactics approval, waiting for VS approval

**Comment by buyaa-n** [comment]
@vatsalyaagrawal @jmarolf is there any update on VS approval? can we merge it?

**Comment by jmarolf** [comment]
@buyaa-n  you are good to merge. This fix has been approved

**Pull request merged**
